src/main/java/net/minecraftforge/datafix/ForgeDataFixerDelegateHandler.java

```diff
@@ -0,0 +1,244 @@
+        //Rebuild of those is not needed.
+        this.activeFixer = new InnerFixer(new Int2ObjectAVLTreeMap<Schema>(schemas), new ArrayList<>(globalList), new IntAVLTreeSet(fixerVersions));
+
+        //This section was added to make it easier to debug the rule set at runtime, since it is no longer pre-computed off thread,
+        //But only on the main thread, at the moment it is needed.
+        //This basically disables the REWRITE_CACHE prefill and fills it on-demand.
+        //Although it is mainly used for debugging, it also provides a good way of reducing DFU's memory footprint, when it is not needed.
+        //Since the following code will compute an update rule from every version to the current for every known type in the schema,
+        //relating to this version. By default, however this is disabled to preserve vanilla compatibility, since it does do the pre-compute
+        //to fill up the cache.
+        //To enable it start the JVM with the system property set to true, -Dforge.datafixer.disablePreCompute=true needs to be added as
+        //a JVM Launch argument.
+        if (Boolean.getBoolean("forge.datafixer.disablePreCompute"))
+        {
+            LOGGER.warn("The pre-compute of the data fixer rules is disabled, this is not recommended for normal use, but can be useful for debugging or in low-memory situations.");
+            return;
+        }
+
+        //We only rebuild once we are told to do so.
+        if (!triggerOffThreadCompute) {
+            return;
+        }
+
+        //Get an iterator for all known versions.
+        final IntBidirectionalIterator iterator = activeFixer.fixerVersions().iterator();
+        while (iterator.hasNext())
+        {
+            //Grab the key.
+            final int versionKey = iterator.nextInt();
+
+            //Grab its schema.
+            final Schema schema = schemas.get(versionKey);
+
+            //For each type in the schema, compute its rewrite rule for the current version.
+            for (final String typeName : schema.types())
+            {
+                //Schedule the recompute off-thread
+                CompletableFuture.runAsync(() -> {
+                    //Grab the type.
+                    final Type<?> dataType = schema.getType(() -> typeName);
+
+                    //Grab its normal rewrite rule for the given version and our current target version.
+                    final TypeRewriteRule rule = activeFixer.getRule(DataFixUtils.getVersion(versionKey), dataVersion);
+
+                    //Optimize the rule to only apply to the given type.
+                    dataType.rewrite(rule, DataFixerUpper.OPTIMIZATION_RULE);
+                }, executor).exceptionally(e -> {
+                    //Fails.
+                    LOGGER.error("Unable to build datafixers", e);
+
+                    //Just kill it, if this fails we have a problem anyway!
+                    Runtime.getRuntime().exit(1);
```

Comment:
I believe we've discussed this already that this is what vanilla does in `DataFixerBuilder`. 

While I'm not so confident on the idea of immediately terminating the JVM for this (because of the possibility of mods causing problems and therefore causing an unfriendly error to users), I will at least require this be noted as a comment to alleviate the concerns of future reviewers and maintainers. Additionally, this can use `System.exit` instead.

Reply:
![image](https://user-images.githubusercontent.com/5585406/157087928-da519e7f-423a-4ebb-8cc1-14d43e2869bc.png)

System.exit is exactly the same anyway....

Reply:
While `Runtime#exit` redirects to `System.exit` internally, it is noted by the javadocs of Runtime that

> The [`System.exit`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/System.html#exit(int)) method is the conventional and convenient means of invoking this method.
> ~ _[`Runtime#exit(int)`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Runtime.html#exit(int)), Java SE 17 & JDK 17_

And while there is credence to the argument of following what vanilla does exactly, there is also the argument that this is akin to the use of, for example, `Lists.newArrayList()` versus `new ArrayList<>()`. However, as this is not too big a deal, I'll leave this point to your choice.